### ISSUE TITLE
Expanding the file/settings/initial midi tab

### ISSUE THREAD

**Issue by davidgritter** [GrandOrgue/grandorgue#1974]
The initial midi tab in grandorgue settings presently allows one to define midi receive event for a limited set of organ elements in a few groups ( manuals, enclosures, sequencer, master controls,  metronome.)  I believe the items in the sequencer group refer to buttons in th combination setter panel.

With increasing use of small displays, custom organ consoles, and headless operation, it would be very useful to expand the number of organ elements that can have an initial midi configuration defined here.  Thus I would suggest that having settings for all the standard panels (those not defined for a specific organ, but present for all organs) would be appreciated.  Specifically, the following additional panels: 

- Crescendo Pedal  (including for the pedal itself)
- Divisionals
- Generals
- Combination setter (partially implemented already)
- Master controls (partially implemented already)
- Recorder/Player
- The coupler panels would be nice but not necessary, and may be difficult since these panels seem to change, depending on the organ loaded

Suggested changes:

- For all elements include the abiltiy to define send events as well as receive events. For example, the Save button iluminates when configuration changes have been made, but not yet saved. Having a send event would allow a button on a remote display to also become illuminated in this case.  Similarly when the combination SET button is active, a remote display or LED might want to know.
- Include all active labels, so that they can send hauptwerk sysex messages to remote displays, allowing display of the loaded organ, current transpose, temperament, memory levels, crescendo level, etc.
- Consider making the initial midi tab list hierarchical  by panel, similar to the audio/midi/organ settings panel, due to the larger number of elements that will be in the list.
- Consider reworking the Master controls panel  to make temperament selection easier.  From the audio/midi/temperament drop down menu in grandorgue, temperament selection is hierarchical.  This could be implemented on the master controls panel by adding category up/category down buttons and a new active label to show the current category. ( this, suggestion, of course, has little to do with the initial midi tab, but with the large number of temperaments, selecting one headlessly is very difficult. Perhaps it might be better to implement some means of assigning midi messages to each available temperament?)

**Comment by willeke1234** [comment]
Maybe it would be more useful to implement #1199 Export and import of MIDI settings in editable format.

**Comment by bonninr** [comment]
I'd like to add a petition for allowing the current loading organ elapsed/remaining time as sysexes to screend.

**Comment by oleg68** [comment]
@larspalo please describe your ideas of improvements of the Initial MIDI settings

**Comment by oleg68** [comment]
@larspalo I'm going to implement your idea in another way.

Recently we've added the `context` column to the MIDI object dialog. So each MIDI object (both internal or odf based) may be referenced by the pair (contect, object name).

My approach is
1. to add the capability of adding more lines to the initial midi dialog and specify the context and the name manually.
2. to implement matching by (contect, object name) with the initial MIDI settings if the object is not configured in the .cmb
3. to rest the old style matching bu `InitialMidiNumber` as well.

**Comment by oleg68** [comment]
I've planned to allow the user to entry the context as string. But another GUI ways are also possible. For example, a new button in the MIDI event editor "To Initial" that copies the current object configuration with the context and the name to the initial midi settings.

**Comment by larspalo** [comment]
> I've planned to allow the user to entry the context as string.

Then it should have some kind of auto complete/correct to avoid spelling mistakes or non usable contexts/elements.

> For example, a new button in the MIDI event editor "To Initial" that copies the current object configuration with the context and the name to the initial midi settings.

Then it's not immediately available from where it's added. I think that a button in the Initial MIDI tab that allow the user to select which line to add from the available contexts/elements would be better.

**Comment by oleg68** [comment]
@larspalo I don't like selecting a line from a listbox with hundreds lines. So I'd prefer to setup an object within the current organ, to check if it works properly and then to store it's settings to initial if I want to share it with other organs.

**Comment by larspalo** [comment]
> So I'd prefer to setup an object within the current organ, to check if it works properly and then to store it's settings to initial if I want to share it with other organs.

@oleg68 That somewhat defeats the very idea of Initial MIDI since it would require first loading/opening an (at least one) organ to make them available to add to Initial MIDI. Also it would make things a lot less comfortable than just adding items from within the same dialog.

> I don't like selecting a line from a listbox with hundreds lines.

I can somewhat understand that, but that's the fact with MIDI-Objects dialog anyway...

What about something like this, perhaps using wxComboBox or wxChoice in the relevant grids:
- allow user to add (new) blank lines with a simple button in Initial MIDI
- click on the context box in that line to select from available contexts
- when a context is selected the elements box will be populated and possible to select from

**Comment by oleg68** [comment]
@larspalo 

> That somewhat defeats the very idea of Initial MIDI since it would require first loading/opening an (at least one) organ to make them available to add to Initial MIDI.

1. From the developer point of view it is hard to implement because now grandorgue creates midi objects only when it is opening an organ. GrandOrgue does not  know what paths exist before opening an organ. It can be reworked but adding new functionality does not compensate for the complexity of the rework.
2. From the user point of view adding new events for MIDI objects without seeing the controls to add events to would require  some kind of abstract thinking from the user. Most of them are musicians and they are not IT specialists. So I think setting up events for a loaded organ and saving parts of them to initial MIDI settings would be more comfortable for most of users.
3. My approach will allow the user to set up not only events for internal MIDI objects but also for ODF-defined ones (if he/she has several ODFs that have similar MIDI objects). Of cause, if the user loads another organ without the objects with paths mentioned in the Initial Setting, they won't be used.

So I'm going
- not to allow the user to  add new objects in the Initial MIDI tab
- instead to add a button `Save to Initial` in the MIDI event dialog

**Comment by larspalo** [comment]
> 
    not to allow the user to add new objects in the Initial MIDI tab
    instead to add a button Save to Initial in the MIDI event dialog

This is a very bad approach. The MIDI-objects dialog has a lot of elements that doesn't even make sense for the user to see (as they cannot be interacted with directly through MIDI). Not to mention that many objects in that dialog are highly specific for one particular organ and should not be possible to set a Initial MIDI configuration for which should be "universal" to any organ.

If you don't want to do the proper work with the Initial MIDI tab I suggest that you postpone any work on this at all and leave it for the future instead.

**Comment by larspalo** [comment]
@oleg68 Sorry, I might have misunderstood your intention a little, but even so, if you might have intended the MIDI event dialog shown when right clicking on an element, this dialog is still available for many elements that should not be considered for adding to Initial MIDI.

**Comment by oleg68** [comment]
> [@oleg68](https://github.com/oleg68) Sorry, I might have misunderstood your intention a little, but even so, if you might have intended the MIDI event dialog shown when right clicking on an element, this dialog is still available for many elements that should not be considered for adding to Initial MIDI.

Yes, I'm talking about the right-click dialog, not about the MIDI Objects dialog.

A user may want to `Save as Initial` the configuration of any object, even an internal or an ODF-defined. But if the object is ODF-defined, a warning message will appear that this object (with certain path) may not exist in other organs. But the user is allowed to enforce saving it: if there is the same path there is another ODF, it may be useful.

**Comment by oleg68** [comment]
My plan of enchancing th Initial MIDI configuration:

1. Add the `Path` column to the Initial MIDI tab
2. Add the matching by path
3. Add the capability to set up `Send` as well as `Receive`
4. Add the 'Save as Initial` button to the right click dialog
5. Add the `Delete` button to the Initial MIDI tab
